### PR TITLE
Feature/vb 1387 new1 return visit times dependent on prisoner location

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Code updates to enable levels based sessions.

## What is the intent behind these changes?

Allows level based sessions to be returned as used in prisons like Bristol.

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/helper/EntityDataLoader.kt [hunk 1]
```diff
@@ -231,3 +255,10 @@ class DeleteEntityHelper(
     prisonRepository.flush()
   }
 }
+
+class AllowedPrisonHierarchy(
+  val level1: String,
```
Comment (AledWEvans) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/helper/EntityDataLoader.kt:260]: called these the same name as in implementation code
levelOneCode
levelTwoCode
levelThreeCode
levelFourCode

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/session/GetSessionsWithLevelsTest.kt [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitRestriction
+import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitStatus
+import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitType
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.Prison
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.SessionTemplate
+import java.time.LocalDate
+import java.time.LocalTime
+
+@DisplayName("Get /visit-sessions")
+class GetSessionsWithLevelsTest(@Autowired private val objectMapper: ObjectMapper) : IntegrationTestBase() {
```
Comment (AledWEvans) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/session/GetSessionsWithLevelsTest.kt:23]: Wow a lot of good test coverage!

**Review by AledWEvans — CHANGES_REQUESTED** [review]
Logic seems good have some requested changes in some naming and use of static methods..

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonerDetailDto.kt [hunk 3]
```diff
@@ -14,5 +14,7 @@ open class PrisonerDetailDto(
   val unitCode2: String?,
 
   @Schema(description = "Level 3 from the prisoner's internal location", example = "008")
-  val unitCode3: String?
+  val unitCode3: String?,
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonerDetailDto.kt:17]: This is something that will be coming from the PrisonAPI call - https://api-dev.prison.service.justice.gov.uk/swagger-ui/index.html?configUrl=/api/api-docs#/prisoners/getPrisonerInformationById. So matching the variable names with the json response. unitCode1 is not (required=true) on the response schema so left it as null.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt [hunk 4]
```diff
@@ -55,13 +57,21 @@ class SessionService(
...
       requestedBookableEndDate
     )
 
+    prisonerId?.let { prisonerIdVal ->
+      prisonerDetailDto = prisonApiService.getPrisonerDetails(prisonerIdVal)
+      prisonerDetailDto?.let { prisonerDetail ->
+        sessionTemplates = sessionTemplates.filter { sessionTemplate -> PrisonerSessionValidator.isSessionAvailableToPrisoner(prisonerDetail, sessionTemplate) }
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt:74]: Agreed, will move into a different method.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonerDetailDto.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonerDetailDto.kt:17]: Bit odd really,... you would have thought they were all in a location, ohwell if thats what there api says... a little worrying though!

[the remark this answers is not in the document]
Comment (rj-adams) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonerDetailDto.kt:17]: This implementation is temporary.

Syscon are currently writing the new endpoint which exposes all 4 levels properly - should be available in dev by the end of this week.

Then at a suitable point we can move over to using this new endpoint.

Their JIRA ticket for this work is [(https://dsdmoj.atlassian.net/browse/SDI-444)]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerSessionValidator.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package uk.gov.justice.digital.hmpps.visitscheduler.utils
+
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.PrisonerDetailDto
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.SessionTemplate
+
+class PrisonerSessionValidator {
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerSessionValidator.kt:6]: Agreed, will incorporate this change.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerLevelMatcher.kt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  override fun test(
+    permittedSessionLocationsList: MutableList<PermittedSessionLocation>?,
+    prisonerDetailDto: PrisonerDetailDto
+  ): Boolean {
+    permittedSessionLocationsList?.let { permittedSessionLocations ->
+      for (permittedSessionLocation in permittedSessionLocations) {
+        val result = LevelMatcher().test(permittedSessionLocation.levelOneCode, prisonerDetailDto.unitCode1)
+          .and(LevelMatcher().test(permittedSessionLocation.levelTwoCode, prisonerDetailDto.unitCode2))
+          .and(LevelMatcher().test(permittedSessionLocation.levelThreeCode, prisonerDetailDto.unitCode3))
+          .and(LevelMatcher().test(permittedSessionLocation.levelFourCode, prisonerDetailDto.unitCode4))
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerLevelMatcher.kt:17]: Agreed, will update code based on below suggestion.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt [hunk 7]
```diff
@@ -55,13 +57,21 @@ class SessionService(
...
       prisonCode,
       requestedBookableStartDate,
       requestedBookableEndDate
     )
 
+    prisonerId?.let { prisonerIdVal ->
+      prisonerDetailDto = prisonApiService.getPrisonerDetails(prisonerIdVal)
+      prisonerDetailDto?.let { prisonerDetail ->
+        sessionTemplates = sessionTemplates.filter { sessionTemplate -> PrisonerSessionValidator.isSessionAvailableToPrisoner(prisonerDetail, sessionTemplate) }
+      }
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt]: Updated

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/helper/EntityDataLoader.kt [hunk 8]
```diff
@@ -165,6 +165,30 @@ class PermittedSessionLocationHelper(
...
+  fun create(sessionTemplate: SessionTemplate, prisonHierarchies: List<AllowedPrisonHierarchy>): MutableList<PermittedSessionLocation> {
+    val permittedSessionLocations = mutableListOf<PermittedSessionLocation>()
+
+    for (prisonHierarchy in prisonHierarchies) {
+      val permittedSessionLocation = repository.saveAndFlush(
+        PermittedSessionLocation(
+          sessionTemplates = mutableListOf(sessionTemplate),
+          prisonId = sessionTemplate.prison.id,
+          prison = sessionTemplate.prison,
+          levelOneCode = prisonHierarchy.level1,
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/helper/EntityDataLoader.kt]: Updated.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerLevelMatcher.kt [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+        if (result)
+          return true
+      }
+    }
+
+    return false
+  }
+
+  inner class LevelMatcher : BiPredicate<String?, String?> {
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerLevelMatcher.kt]: Updated.

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**